webapp/channels/src/packages/mattermost-redux/src/actions/profiles_statuses_groups.ts

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+/* eslint-disable no-console */
+
+import type {PostList, Post} from '@mattermost/types/posts';
+
+import {getProfilesByIds, getStatusesByIds} from 'mattermost-redux/actions/users';
+import {getCurrentChannelId, getCurrentUserId, getIsUserStatusesConfigEnabled} from 'mattermost-redux/selectors/entities/common';
+import type {ActionFunc} from 'mattermost-redux/types/actions';
+
+const pendingUserIdsForProfiles = new Set<string>();
+const MAX_USER_PROFILES_BATCH = 100;
+const USER_PROFILES_DURATION = 5 * 1000;
+let userProfilesIntervalId: NodeJS.Timeout | null = null;
+
+const pendingUserIdsForStatuses = new Set<string>();
+const MAX_USER_STATUSES_BUFFER = 200;
+const USER_STATUSES_DURATION = 5 * 1000;
+const USER_STATUSES_REQUEST_SURGE_THRESHOLD = 1000;
+let userStatusesIntervalId: NodeJS.Timeout | null = null;
+let haveUserStatusRequestsSurged = false;
+
+function addUserIdForProfiles(userId: string): ActionFunc<boolean> {
+    return (dispatch) => {
+        function getPendingProfilesById() {
+            console.log('addUserIdForProfiles start', pendingUserIdsForProfiles);
+
+            dispatch(getProfilesByIds(Array.from(pendingUserIdsForProfiles)));
+            pendingUserIdsForProfiles.clear();
```

Comment:
This won't matter in the current version of this code, but to be safe, could you move this to earlier in the function? If we were to ever add an await before this line in the function, it could cause this to be interrupted and either for the function to run again with the same `pendingUserIdsForProfiles` or a caller could perhaps add an extra ID to it after we'd made the request but before the await resolved, causing the extra ID to be deleted without being fetched.

If this were a language with actual multi-threading, you'd typically use some sort of atomic action to get the value of the set and then clear it in one action to prevent another thread from modifying it before we make the request. Since this is JS where function execution can only be interrupted by something like an await, I think it would be safe enough to start this function with something like
```js
const ids = Array.from(pendingUserIdsForProfiles);
pendingUserIdsForProfiles.clear();
```
That isn't foolproof, and we would still run into issues if someone added an `await` between those, but by doing those first, it should lower the chance that that happens

Reply:
yes extracted it from set to array before